### PR TITLE
Extended forward() for using in closure

### PR DESCRIPTION
**Pull request by PhucVH888**
### Introduction
To recall, the `forward` is designed to delegate the future and to be fulfilled by other delegated method. In summary, `forward` can be used in the following four cases.

1. DONE: forward MethodCall (#682)
2. DONE: forward FutureChaining (#784)
3. DONE: forward PreexistingFuture (#784)
4. forward in closure's body

In this PR, the `forward()` can be used in closure.

### Syntax
```
4. forward(arg ~~> fun () => forward(tmp)
```
#### 4. `forward in closure's body`
```
      forward(f ~~> fun(x:int) => forward(g))
```
where `f` and `g` are preexisting futures.
```
class Base
  def one(g : Fut[int]) : int
    forward( (new Foo) ! foo() ~~> fun(x : int) => forward(g))
  end
end
```
### Examples
You can find many examples in the src/tests/encore/forward folder.

### CODE REVIEW

File: src/tests/encore/forward/forwardMethodCallInClosureBody.enc [hunk 1]
```diff
@@ -0,0 +1,18 @@
+active class Base
+  def base() : int
+    42
+  end
+end
+
+active class Foo
+  def foo(arg : Fut[int]) : int
+    forward(arg ~~> fun(x : int) : int => forward((new Base) ! base()))
```
Comment (supercooldave) [inline src/tests/encore/forward/forwardMethodCallInClosureBody.enc]: Why double forward in all the tests?

**Comment by supercooldave** [comment]
Here is the basis for the key test case:
```
fun join[t](ff : Fut[Fut[t]]) : Fut[t]
  ff ~~> (fun (f : Fut[t]) forward(f))
end
```
This is the function that asynchronously flattens a future of futures.

Could be used like:
```
active class Foo
  def m() : int
    10
  end

  def bar() : Fut[int]
    this!m()
  end
end

active class Main
  def main() : int
    val foo = new Foo
    val Fut[t] f = join(foo!bar())
    println("{}", get(f))
  end
end
```
Make it so.

**Comment by supercooldave** [comment]
Here is another test case to consider (it can probably be simplified):
```

active class Main
  def main() : unit
    val fres = this!foo() ~~> (fun (x : int) : int forward(this!inc(x)) end))
    print({}, get(fres))
  end

  def foo() : int
    10
  end
  
  def inc(x : int) : int
    x + 1
  end
end
```
The point of this is that the return type of the method within which `forward` appears is different from the return type of the closure `forward` appears within.

A similar example should be devised for `forward` inside `async`.

File: src/back/CodeGen/Expr.hs [hunk 2]
```diff
@@ -1136,27 +1156,47 @@ instance Translatable A.Expr (State Ctx.Context (CCode Lval, CCode Stat)) where
...
 
   translate clos@(A.Closure{A.eparams, A.body}) = do
     tmp <- Ctx.genSym
+    fut <- Ctx.genNamedSym "fut"
     globalFunctionNames <- gets Ctx.getGlobalFunctionNames
+    isAsyncForward <- gets Ctx.isAsyncForward
     let bound = map (ID.qLocal . A.pname) eparams
         freeVars = filter (ID.isLocalQName . fst) $
                    Util.freeVariables bound body
+        isIdClosure = not . null $ filter (isIdFun . A.pname) eparams
```
Comment (supercooldave) [inline src/back/CodeGen/Expr.hs]: Why are you checking for the identity function?
This suggests to me that the design is broken somewhere.

**Review by supercooldave — COMMENTED** [review]
Some small comments.

**Comment by supercooldave** [comment]
The following case does not compile (when used in a larger context):
```
fun join(ff : Fut[Fut[int]]) : Fut[int]
  forward(ff ~~> (fun (f : Fut[int]) : int => forward(f)))
end
```

Error is:
```
 *** Error during typechecking *** 
"join.enc" (line 2, column 14)
Forward cannot be used in functions
```

But this is a forward inside a closure inside a future chain, which **is** valid.

**Comment by supercooldave** [comment]
If I modify `forwardArgInClosure.enc` so that `join` is polymorphic, as follows:
```
def join[sharable t](ff : Fut[Fut[t]]) : t
  forward(ff ~~> (fun (f : Fut[t]) : t => forward(f)))
end
```
I get the error
```
 *** Error during typechecking *** 
"join.enc" (line 9, column 53)
Returned type t of forward should match with the result type of the containing method Fut[sharable t]
In expression: 
```
which suggests that something is wrong in the type checker.

**Comment by supercooldave** [comment]
It seems that I get a similar error if I try to use `forward` inside a chained-closure where the return type is an object type. 

For instance,
```
active class Main
  def main() : unit
    val fres = (this!foo() ~~>  fun (x : int) : int
                                    val bar = new Bar
                                    forward(bar!bar())
                                end)
    await(fres)
    println("{}", get(fres))
  end

  def foo() : int
    42
  end
end

active class Bar  
  def bar() : int
    10
  end
end
```
will succeed, but if I change the `Bar!bar()` method to return something of some other type, such as a passive object, it fails. For instance, the following:
```
active class Main
  def main() : unit
    val fres = (this!foo() ~~>  fun (x : int) : String
                                    val bar = new Bar
                                    forward(bar!bar())
                                end)
    await(fres)
    println("{}", (get(fres)))
  end

  def foo() : int
    42
  end
end

active class Bar  
  def bar() : String
    "Bar"
  end
end
```
results in error:
```
 *** Error during typechecking *** 
"dave.enc" (line 5, column 48)
Returned type String of forward should match with the result type of the containing method Fut[String.String]
```

**Comment by PhucVH888** [comment]
@supercooldave I have fixed the issues and included your test cases. This PR is ready for review, IMO.

**Comment by supercooldave** [comment]
I'll have a look.
Did you handle the case of `forward` in `async`?

**Comment by supercooldave** [comment]
Have you addressed issue  #816?

File: src/back/CodeGen/Expr.hs [hunk 3]
```diff
@@ -1019,26 +1020,38 @@ instance Translatable A.Expr (State Ctx.Context (CCode Lval, CCode Stat)) where
         callTheMethodOneway
           ntarget targetType name args typeArguments Ty.unitType
 
-      let nullCheck = targetNullCheck ntarget target name emeta "."
+      let nullCheck = targetNullCheck ntarget target name emeta "!"
```
Comment (supercooldave) [inline src/back/CodeGen/Expr.hs]: Why is this changed from "." to "!"?

File: src/tests/encore/forward/failedPolyTypeStreamWithObject.enc [hunk 4]
```diff
@@ -0,0 +1,58 @@
+-- Error related to unknown _enc_type_t in method produce()
```
Comment (supercooldave) [inline src/tests/encore/forward/failedPolyTypeStreamWithObject.enc]: clean up the comments in these files = remove them.

File: src/tests/encore/forward/test0613.enc [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+  def green(arg : Fut[String]) : String
+    (arg ~~> fun(x : String) : unit
+                    "jdfdf"
+                    forward((new Crop(x)) ! collect())
+             end)
+    -- forward((new Crop("Never mind")) ! collect())
+    "This message should never be appeared when calling method green()"
+  end
+end
+-- active class Pepper
```
Comment (supercooldave) [inline src/tests/encore/forward/test0613.enc]: Remove commented out code, please.

**Review by supercooldave — COMMENTED** [review]
A first batch of comments.

**Comment by supercooldave** [comment]
The following test cases don't have a `.out`:
```
failedPolyTypeStreamWithObject.enc
forwardReturn.enc
test0613.enc
```

**Comment by supercooldave** [comment]
The follow code should print 10 three times, but it prints 10 only once. 
```

fun join[t](ff : Fut[Fut[t]]) : Fut[t]
  ff ~~> (fun (f : Fut[t]) : t => forward(f))
end

fun intJoin(ff : Fut[Fut[int]]) : Fut[int]
  ff ~~> (fun (f : Fut[int]) : int => forward(f))
end

active class Bar
  def bar() : int
    10
  end
  
  def baz() : Fut[int]
    this!bar()
  end
end

active class Main
  def main() : unit
    val bar = new Bar()
    println("{}", get(bar!bar()))
    println("{}", get(intJoin(bar!baz())))
    println("{}", get(join(bar!baz())))
  end
end
```

File: src/back/CodeGen/Expr.hs (hunk 3 above)
Comment (PhucVH888) [reply src/back/CodeGen/Expr.hs]: "!" is used for my own test, but I forgot to recover before pushing the code. It has been fixed in the next commit.

File: src/tests/encore/forward/failedPolyTypeStreamWithObject.enc (hunk 4 above)
Comment (PhucVH888) [reply src/tests/encore/forward/failedPolyTypeStreamWithObject.enc]: This test is irrelevant to the forward. It's accidentally committed.

File: src/tests/encore/forward/test0613.enc (hunk 5 above)
Comment (PhucVH888) [reply src/tests/encore/forward/test0613.enc]: This test is for debugging, be removed. Thanks Dave.

**Comment by PhucVH888** [comment]
@supercooldave : The `forward` is not allowed to be used in function, it has been checked in the test case `forwardFunction.enc`. However, the case that `forward` is nested in the `closure` is not rejected. I will fixed it.

**Comment by PhucVH888** [comment]
The fix of issue #816 (Forward-Return) has been included in this PR.

**Comment by supercooldave** [comment]
Are you saying that you won't allow the following?
```
fun join[t](ff : Fut[Fut[t]]) : Fut[t]
   ff ~~> (fun (f : Fut[t]) : t => forward(f))
end
 ```
The test seems to suggest that this is prohibited.
But it should be allowed. Indeed, it is one of the core use case for `forward`.

If you look at the semantics of `forward` we were working on, you'll see that this is completely sensible. Firstly, the `forward` gets translated as follows:
```
fun join[t](ff : Fut[Fut[t]]) : Fut[t]
   ff ~~> (fun (f : Fut[t]) : t => forward(f ~~> id) )
end
 ```
Now the semantics are that the future associated with the outer `~~>` is used to catch the result of the inner `~~>`.

File: src/tests/encore/forward/forwardFunctionNested.enc [hunk 6]
```diff
@@ -0,0 +1,26 @@
+fun join[t](ff : Fut[Fut[t]]) : Fut[t]
+  ff ~~> (fun (f : Fut[t]) : t => forward(f))
+end
```
Comment (supercooldave) [inline src/tests/encore/forward/forwardFunctionNested.enc:3]: This should be allowed.

File: src/tests/encore/forward/forwardFunctionNested.enc [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+
+fun intJoin(ff : Fut[Fut[int]]) : Fut[int]
+  ff ~~> (fun (f : Fut[int]) : int => forward(f))
+end
```
Comment (supercooldave) [inline src/tests/encore/forward/forwardFunctionNested.enc:7]: This should be allowed.

**Review by supercooldave — COMMENTED** [review]
Forward inside chaining or inside `async` should always work, regardless of where the forwarding or chaining occurs.

**Comment by supercooldave** [comment]
It's unclear whether this will be finished.
For the limited scope that this version covers, it does what it should and the code has been reviewed and comments mostly addressed adequately.
I feel it should be accepted before bit rot sets in.

**Comment by kikofernandez** [comment]
Does this mean that you will maintain `forward` in Encore?

**Comment by supercooldave** [comment]
In the absence of other maintainers, yes.
There are one or two variants that were not implemented that need to be done as well.

**Comment by kikofernandez** [comment]
I am interested in `forward` and its use with futures and ParTs. I cannot commit to maintain it in the long run (paternity leave) but, if there is a hackathon soon, we should clarify the remaining work and finish it up

**Comment by supercooldave** [comment]
Forward is useful for Encore.
And I think a hackathon should be enough to finish the remaining work.

Don't worry @kikofernandez, I won't ask you to maintain this.
